stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift

```diff
@@ -15,10 +15,133 @@ import SwiftPrivate
 import Darwin
 #elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
 import Glibc
+#elseif os(Windows)
+import MSVCRT
+import WinSDK
 #endif
 
-
+internal func _signalToString(_ signal: Int) -> String {
+  switch CInt(signal) {
+  case SIGILL:  return "SIGILL"
+  case SIGABRT: return "SIGABRT"
+  case SIGFPE:  return "SIGFPE"
+  case SIGSEGV: return "SIGSEGV"
 #if !os(Windows)
+  case SIGTRAP: return "SIGTRAP"
+  case SIGBUS:  return "SIGBUS"
+  case SIGSYS:  return "SIGSYS"
+#endif
+  default:      return "SIG???? (\(signal))"
+  }
+}
+
+public enum ProcessTerminationStatus : CustomStringConvertible {
+  case exit(Int)
+  case signal(Int)
+
+  public var description: String {
+    switch self {
+    case .exit(let status):
+      return "Exit(\(status))"
+    case .signal(let signal):
+      return "Signal(\(_signalToString(signal)))"
+    }
+  }
+}
+
+
+#if os(Windows)
+public func spawnChild(_ args: [String])
+    -> (process: HANDLE, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  var _stdin: (read: HANDLE?, write: HANDLE?)
+  var _stdout: (read: HANDLE?, write: HANDLE?)
+  var _stderr: (read: HANDLE?, write: HANDLE?)
+
+  var saAttributes: SECURITY_ATTRIBUTES = SECURITY_ATTRIBUTES()
+  saAttributes.nLength = DWORD(MemoryLayout<SECURITY_ATTRIBUTES>.size)
+  saAttributes.bInheritHandle = TRUE
+  saAttributes.lpSecurityDescriptor = nil
+
+  if CreatePipe(&_stdin.read, &_stdin.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stdin.write, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  if CreatePipe(&_stdout.read, &_stdout.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stdout.read, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  if CreatePipe(&_stderr.read, &_stderr.write, &saAttributes, 0) == FALSE {
+    fatalError("CreatePipe() failed")
+  }
+  if SetHandleInformation(_stderr.read, HANDLE_FLAG_INHERIT, 0) == FALSE {
+    fatalError("SetHandleInformation() failed")
+  }
+
+  var siStartupInfo: STARTUPINFO = STARTUPINFO()
+  siStartupInfo.cb = DWORD(MemoryLayout<STARTUPINFO>.size)
+  siStartupInfo.hStdError = _stderr.write
+  siStartupInfo.hStdOutput = _stdout.write
+  siStartupInfo.hStdInput = _stdin.read
+  siStartupInfo.dwFlags |= STARTF_USESTDHANDLES
+
+  var piProcessInfo: PROCESS_INFORMATION = PROCESS_INFORMATION()
+
+  if CreateProcessA(args.joined(separator: " "), nil, nil, nil, TRUE,
```

Comment:
Why CreateProcessA and not CreateProcessW? This is not Unicode safe (the *A APIs are *NOT* UTF-8)...

Also, please add a comment "// FIXME Properly quote arguments" if you don't plan to do it in the initial patch; you're going to need to do some *very* complex escaping of the arguments here. Here is at least one good resource: https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/

Reply:
The `W` variant is UTF-16, and requires conversion, which can cause an allocation (this is pretty hot during the testing phase).  But, I can switch to the `W` variant.  Sure, I'll add a FIXME for that, I want to get through as many of the tests as possible :-).

Reply:
Using CreateProcessW is absolutely required. As I said, CreateProcessA does not support Unicode at all[1]. If you pass a UTF-8 Unicode string to CreateProcessA, it'll just get interpreted as the current Windows 8-bit encoding and become garbage data. So attempting to launch a process with Unicode characters in the filename would simply fail. Unfortunately, you can't avoid the allocation.

The same goes for *any* Windows APIs used throughout Swift, we should never ever be using anything *A, and be using exclusively the *W family of APIs.

[1] Actually, very recent versions of Windows (late 2017 update to Windows 10) allow setting UTF-8 as the system code page, meaning that standard C APIs and *A functions *can* take UTF-8, but because that is not the default, can be changed by the user to any other code page, and will not work at all on any version of Windows prior to that recent update, we cannot rely on it.

Reply:
Alright, I think I switched the implementation correctly to `CreateProcessW`.  If you spot anything else that is wrong, please do let me know.  There are still a couple of holes that need to be addressed, like waiting on the child's pipes.